File: app/src/main/java/com/example/alarm/AlarmsAdapter.java
package com.example.alarm;

import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.alarm.databinding.ItemAlarmBinding;
import com.example.alarm.model.Alarm;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class AlarmsAdapter extends RecyclerView.Adapter<AlarmsAdapter.AlarmsViewHolder>
        implements View.OnClickListener, View.OnLongClickListener {

    private final int interval;
    private final AlarmActionListener actionListener;

    List<Alarm> alarms = new ArrayList<>();
    boolean canLongClick = true;
    private final Set<Integer> checkedPositions = new HashSet<>();
    private int index = 0;

    public AlarmsAdapter(int interval, AlarmActionListener actionListener) {
        this.interval = interval;
        this.actionListener = actionListener;
    }

    public void setAlarms(List<Alarm> alarms) {
        this.alarms = alarms;
        notifyDataSetChanged();
    }

    public List<Alarm> getDeleteList() {
        List<Alarm> list = new ArrayList<>();
        for (int i : checkedPositions) {
            list.add(alarms.get(i));
        }
        return list;
    }

    public void clearPositions() {
        canLongClick = true;
        checkedPositions.clear();
    }

    private void savePosition(Alarm alarm) {
        for (int i = 0; i < alarms.size(); i++) {
            if (alarms.get(i).equals(alarm)) {
                if (checkedPositions.contains(i)) {
                    checkedPositions.remove(i);
                } else {
                    checkedPositions.add(i);
                }
                break;
            }
        }
    }

    @Override
    public void onClick(View v) {
        Alarm alarm = (Alarm) v.getTag();
        if(v.getId() == R.id.switch1) {
            if (canLongClick) {
                for (int i = 0; i < alarms.size(); i++) {
                    if (alarms.get(i).equals(alarm)) {
                        index = i;
                        break;
                    }
                }
                actionListener.onAlarmEnabled(alarm, index);
            }
        } else if (v.getId() == R.id.checkBox) {
            if (!canLongClick) {
                savePosition(alarm);
            }
        } else {
            if (canLongClick) {
                actionListener.onAlarmChange(alarm);
            } else {
                savePosition(alarm);
                notifyDataSetChanged();
            }
        }
    }

    @Override
    public boolean onLongClick(View v) {
        if (canLongClick) {
            Alarm alarm = (Alarm) v.getTag();
            savePosition(alarm);
            actionListener.onAlarmLongClicked();
            notifyDataSetChanged();
            canLongClick = false;
        }
        return true;
    }

    @NonNull
    @Override
    public AlarmsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        ItemAlarmBinding binding = ItemAlarmBinding.inflate(inflater, parent, false);
        binding.getRoot().setOnClickListener(this);
        binding.getRoot().setOnLongClickListener(this);
        binding.switch1.setOnClickListener(this);
        binding.checkBox.setOnClickListener(this);
        return new AlarmsViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull AlarmsViewHolder holder, int position) {
        Alarm alarm = alarms.get(position);
        holder.itemView.setTag(alarm);
        holder.binding.switch1.setTag(alarm);

        String tm = (alarm.getTimeMinutes() == 0) ? "0" : "";
        String tm2 = (alarm.getTimeMinutes() >= 1 && alarm.getTimeMinutes() <= 9) ? "0" : "";

        String timeText = alarm.getTimeHours() + ":" + tm2 + alarm.getTimeMinutes() + tm;
        holder.binding.timeTextView.setText(timeText);

        String intervalText = (alarm.getName().equals("default"))
                ? "раз в " + interval + " минут"
                : "<font color='#FF00FF'>" + alarm.getName() + "</font>";

        if (interval == 3 && alarm.getName().equals("default")) {
            intervalText += "ы";
        }

        holder.binding.intervalTextView.setText(Html.fromHtml(intervalText, Html.FROM_HTML_MODE_LEGACY));
        holder.binding.switch1.setChecked(alarm.getEnabled());

        if (!canLongClick) {
            holder.binding.checkBox.setTag(alarm);
            holder.binding.switch1.setVisibility(View.INVISIBLE);
            holder.binding.checkBox.setVisibility(View.VISIBLE);
            holder.binding.checkBox.setChecked(checkedPositions.contains(position));
        } else {
            holder.binding.switch1.setVisibility(View.VISIBLE);
            holder.binding.checkBox.setVisibility(View.GONE);
        }
    }

    @Override
    public int getItemCount() {
        return alarms.size();
    }

    public static class AlarmsViewHolder extends RecyclerView.ViewHolder {
        final ItemAlarmBinding binding;

        public AlarmsViewHolder(ItemAlarmBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/example/alarm/AlarmsAdapter.kt
package com.example.alarm

import android.annotation.SuppressLint
import android.text.Html
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.example.alarm.databinding.ItemAlarmBinding
import com.example.alarm.model.Alarm
import com.example.alarm.model.AlarmService
import com.example.alarm.model.Settings
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

interface AlarmActionListener {
    fun onAlarmEnabled(alarm: Alarm, index: Int)
    fun onAlarmChange(alarm: Alarm)
    fun onAlarmLongClicked()
}

class AlarmsAdapter(
    private val settings: Settings,
    private val actionListener: AlarmActionListener
) : RecyclerView.Adapter<AlarmsAdapter.AlarmsViewHolder>(), View.OnClickListener, View.OnLongClickListener {

    var alarms: List<Alarm> = emptyList()
        @SuppressLint("NotifyDataSetChanged")
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    var canLongClick: Boolean = true
    private var checkedPositions: MutableSet<Int> = mutableSetOf()
    private var index = 0

    fun getDeleteList(): List<Alarm> {
        val list = mutableListOf<Alarm>()
        for(i in checkedPositions) list.add(alarms[i])
        return list
    }

    fun clearPositions() {
        canLongClick = true
        checkedPositions.clear()
    }
    private fun savePosition(alarm: Alarm) {
            for (i in alarms.indices) {
                if (alarms[i] == alarm) {
                    if (i in checkedPositions) {
                        checkedPositions.remove(i)
                    } else {
                        checkedPositions.add(i)
                    }
                    break
                }
            }
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun onClick(v: View) {
        val alarm = v.tag as Alarm
        when(v.id) {
            R.id.switch1 -> {
                if(canLongClick) {
                    for (i in alarms.indices) {
                        if(alarms[i] == alarm) {
                            index = i
                            break
                        }
                    }
                    actionListener.onAlarmEnabled(alarm, index)
                }
            }
            R.id.checkBox -> {
                if(!canLongClick) {
                    savePosition(alarm)
                }
            }
            else -> {
                if(canLongClick) actionListener.onAlarmChange(alarm)
                else {
                    savePosition(alarm)
                    notifyDataSetChanged()
                }
            }
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun onLongClick(v: View?): Boolean {
        if(canLongClick) {
            val alarm = v?.tag as Alarm
            savePosition(alarm)
            actionListener.onAlarmLongClicked()
            notifyDataSetChanged()
            canLongClick = false
        }
        return true
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AlarmsViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = ItemAlarmBinding.inflate(inflater, parent, false)
        binding.root.setOnClickListener(this) //list<alarm> element
        binding.root.setOnLongClickListener(this)
        binding.switch1.setOnClickListener(this)
        binding.checkBox.setOnClickListener(this)
        binding.root.isHapticFeedbackEnabled = true //for test
        return AlarmsViewHolder(binding)
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun onBindViewHolder(holder: AlarmsViewHolder, position: Int) {
            val alarm = alarms[position]
            with(holder.binding) {
                holder.itemView.tag = alarm
                switch1.tag = alarm
                var tm = ""
                var tm2 = ""
                if (alarm.timeMinutes == 0) tm = "0"
                if (alarm.timeMinutes in 1..9) tm2 = "0"
                val txt: String = "${alarm.timeHours}:${tm2}${alarm.timeMinutes}${tm}"
                timeTextView.text = txt
                var txt2: String = ""
                txt2 += if (alarm.name != "default")
                    "<font color='#FF00FF'>${alarm.name}</font>"
                else "раз в ${settings.interval} минут"
                if (settings.interval == 3 && alarm.name == "default") txt2 += "ы"
                intervalTextView.text = Html.fromHtml(txt2, 0)
                switch1.isChecked = alarm.enabled == 1
                if (!canLongClick) {
                    checkBox.tag = alarm
                    switch1.visibility = View.INVISIBLE
                    checkBox.visibility = View.VISIBLE
                    checkBox.isChecked = position in checkedPositions
                } else {
                    switch1.visibility = View.VISIBLE
                    checkBox.visibility = View.GONE
                }
            }
    }
    override fun getItemCount(): Int = alarms.size

    class AlarmsViewHolder(
        val binding: ItemAlarmBinding
    ) : RecyclerView.ViewHolder(binding.root)
}